src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
-const doesThisCodeHaveErrors = "not if I comment it out";
-export default doesThisCodeHaveErrors;
+import Block, { BlockBase, KnownBlock, ProcedureBlock } from "../../Block";
+import Costume from "../../Costume";
+import Project, { TextToSpeechLanguage } from "../../Project";
+import Script from "../../Script";
+import Sound from "../../Sound";
+import Target, { Sprite, Stage } from "../../Target";
+import * as BlockInput from "../../BlockInput";
+import * as sb3 from "./interfaces";
+import { OpCode } from "../../OpCode";
+import { generateId } from "../../util/id";
+import { prop } from "../../util/ts-util";
 
-/*
-import * as JSZip from "jszip";
-import * as sb3 from "../sb3";
+const BIS = sb3.BlockInputStatus;
 
-import Project from "../../Project";
-import Target from "../../Target";
+interface ToSb3Options {
+  warn: (message: string) => void;
+}
+
+interface ToSb3Output {
+  json: string;
+}
 
-import List from "../../List";
-import Variable from "../../Variable";
+export default function toSb3(options: Partial<ToSb3Options> = {}): ToSb3Output {
+  // Serialize a project. Returns an object containing the text to be stored
+  // in the caller's project.json output file.
 
-export default async function toSb3(project: Project): Promise<ArrayBuffer> {
-  const outZip = new JSZip();
+  const project: Project = this;
 
-  for (const target of [project.stage, ...project.sprites]) {
-    for (const costume of target.costumes) {
-      outZip.file(`${costume.md5}.${costume.ext}`, costume.asset);
+  let warn = (message: string) => undefined;
+  if (options.warn) {
+    warn = options.warn;
+  }
+
+  type BlockData = sb3.Target["blocks"];
+
+  function applyBlockData(a: BlockData, b?: BlockData) {
+    if (b) {
+      Object.assign(a, b);
     }
   }
 
-  outZip.file("project.json", JSON.stringify(getProjectJSON(project)));
+  function newBlockData(): BlockData {
+    return {};
```

Comment:
What purpose does this function serve?

Reply:
If this is about functions not needing to know `BlockData`'s internal data structure, why?

`BlockData` matches the `blocks` section of a `.sb3` Target, so unless the Scratch Team decides to make `blocks` an `Array` in all future `.sb3`s, this abstraction seems unnecessary to me: the "internal" data structure isn't actually so internal--it's an essential property of the `.sb3` format itself.

Reply:
You can still capture the type/semantics with
```ts
const blockData: BlockData = {};
```